File: src/mersenne_prime.rs
use indicatif::{ProgressBar, ProgressStyle};
use rug::Integer;

pub(crate) fn is_mersenne_prime(number: u64) -> bool {
    let mut sequence_value: Integer = Integer::from(4);
    let mut mersenne_candidate: Integer = Integer::from(1);
    mersenne_candidate <<= number as u32; // 2^number
    mersenne_candidate -= 1; // 2^number - 1

    let iterations: u64 = number - 2;
    let progress_bar: ProgressBar = ProgressBar::new(iterations);
    progress_bar.set_style(
        ProgressStyle::with_template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} ({eta})")
            .unwrap()
            .progress_chars("=> "),
    );

    for _ in 0..iterations {
        sequence_value.square_mut();
        sequence_value -= 2;
        sequence_value %= &mersenne_candidate;
        progress_bar.inc(1);
    }
    progress_bar.finish_with_message("Lucas-Lehmer test complete");

    sequence_value == 0
}

File: src/MersennePrime.java
import java.math.BigInteger;

public class MersennePrime {
    public static boolean is_mersenne_prime(int number) {
        BigInteger sequence_value = new BigInteger("4");
        BigInteger mersenne_candidate = new BigInteger("1");
        mersenne_candidate = mersenne_candidate.multiply(BigInteger.TWO.pow(number)).subtract(BigInteger.ONE);

        int iterations = number - 2;

        for (int i = 0; i < iterations; i++) {
            sequence_value = sequence_value.multiply(sequence_value);
            sequence_value = sequence_value.subtract(BigInteger.TWO);
            sequence_value = sequence_value.mod(mersenne_candidate);
            System.out.printf("%d/%d%n", i + 1, iterations);
        }
        System.out.println("Lucas-Lehmer test complete");

        return sequence_value.equals(BigInteger.ZERO);
    }
}
